sdk/python/lib/test/automation/test_remote_workspace.py

```diff
@@ -12,10 +12,111 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import os
+from typing import Optional
 import pytest
 
 from pulumi.automation._remote_workspace import _is_fully_qualified_stack_name
 
+from pulumi.automation import (
+    LocalWorkspace,
+    OpType,
+    RemoteGitAuth,
+    RemoteWorkspaceOptions,
+    create_remote_stack_git_source,
+    create_or_select_remote_stack_git_source,
+    select_remote_stack_git_source,
+)
+
+from .test_utils import stack_namer
+
+
+test_repo = "https://github.com/pulumi/test-repo.git"
+
+
+@pytest.mark.parametrize("factory", [
+    create_remote_stack_git_source,
+    create_or_select_remote_stack_git_source,
+    select_remote_stack_git_source,
+])
+@pytest.mark.parametrize("error,stack_name,url,branch,commit_hash,auth", [
+    ('stack name "" must be fully qualified.', "", "", None, None, None),
+    ('stack name "name" must be fully qualified.', "name", "", None, None, None),
+    ('stack name "owner/name" must be fully qualified.', "owner/name", "", None, None, None),
+    ('stack name "/" must be fully qualified.', "/", "", None, None, None),
+    ('stack name "//" must be fully qualified.', "//", "", None, None, None),
+    ('stack name "///" must be fully qualified.', "///", "", None, None, None),
+    ('stack name "owner/project/stack/wat" must be fully qualified.', "owner/project/stack/wat", "", None, None, None),
+    ('url is required.', "owner/project/stack", None, None, None, None),
+    ('url is required.', "owner/project/stack", "", None, None, None),
+    ('either branch or commit_hash is required.', "owner/project/stack", test_repo, None, None, None),
+    ('either branch or commit_hash is required.', "owner/project/stack", test_repo, "", "", None),
+    ('branch and commit_hash cannot both be specified.', "owner/project/stack", test_repo, "branch", "commit", None),
+    ('ssh_private_key and ssh_private_key_path cannot both be specified.', "owner/project/stack", test_repo, "branch",
+        None, RemoteGitAuth(ssh_private_key="key", ssh_private_key_path="path")),
+])
+def test_remote_workspace_errors(
+    factory,
+    error: str,
+    stack_name: str,
+    url: str,
+    branch: Optional[str],
+    commit_hash: Optional[str],
+    auth: Optional[RemoteGitAuth],
+):
+    with pytest.raises(Exception) as e_info:
+        factory(stack_name=stack_name, url=url, branch=branch, commit_hash=commit_hash, auth=auth)
+    assert str(e_info.value) == error
+
+
+@pytest.mark.parametrize("factory", [
+    create_remote_stack_git_source,
+    create_or_select_remote_stack_git_source,
+])
+@pytest.mark.skipif("PULUMI_ACCESS_TOKEN" not in os.environ, reason="PULUMI_ACCESS_TOKEN not set")
+def test_remote_workspace_stack_lifecycle(factory):
```

Comment:
Unfortunately, this test takes ~3 minutes (for each factory, so total ~6 minutes). What can we do to avoid running it in PRs?

Reply:
We'll disable the test by default (requires `PULUMI_TEST_DEPLOYMENTS_API` to be set). And consider running the test in a cron until we add support for mocking the deployments api.